sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala

```diff
@@ -658,6 +658,41 @@ abstract class CastBase extends UnaryExpression with TimeZoneAwareExpression wit
     }
   }
 
+  private[this] def stringToDecimal(str: UTF8String, decimalType: DecimalType): Decimal = {
+    val bigDecimal = try {
+      // According the benchmark test,  `s.toString.trim` is much faster than `s.trim.toString`.
```

Comment:
it sounds like `UTF8String.trim` is not implemented well. We may need to look at how `java.lang.String.trim` is implemented.

Reply:
Filed a ticket: https://issues.apache.org/jira/browse/SPARK-32891